File: leet/0720_pre/leet_633.py
# check whether an integer z is a sum of two squares: z = x^2 + y^2
import math
class Solution(object):
    def judgeSquareSum(self, c):
        '''
        :type c: int
        :rtype: bool
        '''
        mid = int(math.sqrt(c))
        found = False
        for i in range(mid, -1, -1):
            if isSquareNum(c - i * i):
                found = True
                break
        return found

def isSquareNum(n):
    '''
    :type n: int
    :rtype: bool
    '''
    tmp = int(math.sqrt(n))
    return (n == tmp * tmp)

if __name__ == '__main__':
    solu = Solution()
    print solu.judgeSquareSum(3) 

File: leet/0720_pre/leet_633.cpp
#include <iostream>
#include <cmath>
using namespace std;
class Solution {
public:
    bool judgeSquareSum(int c) {
	int mid = (int)sqrt(c);
	bool found = false;
	for(int i = mid; i >= 0; --i)
	{
	    if(isSquareNum(c - i * i)) // find x^2
	    {
		found = true;
		break;
	    }
	}
	return found;
    }
    bool isSquareNum(int n)
    {
        int tmp = (int)sqrt(n);
        return (n == tmp * tmp);
    }
};

int main()
{
    Solution solu;
    cout << solu.judgeSquareSum(0) << endl;
    return 0;
}
